source/kernel/src/bbq.rs

```diff
@@ -84,14 +103,166 @@ impl Drop for BBQStorage {
     }
 }
 
-impl BBQBidiHandle {
+use abi::bbqueue_ipc::{
+    GrantR as InnerGrantR,
+    GrantW as InnerGrantW,
+};
+
+pub struct GrantW {
+    grant: InnerGrantW<'static>,
+    // When we commit, we trigger OUR commit waitcell
+    our_commit: NonNull<WaitCell>,
+}
```

Comment:
these `Grant` types need to keep the `BBQStorage`'s ref count alive, right? how is that ensured exactly?

Reply:
It's not! That's a good call. In the "real" version of bbqueue, grants inherit the lifetime param from their producer/consumer, which derive it from the lifetime of the backing storage, which 99% of the time is `'static`. When I hacked this all up for the mnemos version of bbqueue, that kind of went out the window.

I *think* I want to kick this can down the road, but I would like to open an issue to make sure this gets fixed. I'm not sure whether I want the grants holding an actual heap handle, or add a manual incr/decr refcount on an arc.

Reply:
Ehhh, just changed my mind, I'll go ahead and add the Arc here, it should just be a pointer.